packages/mui-system/src/createStyled.js

```diff
@@ -217,105 +153,52 @@ export default function createStyled(input = {}) {
-                    props,
-                    transformVariants(styledArgVariants),
-                    styledArgVariants,
-                  );
-                  variantStyles.forEach((variantStyle) => {
-                    result = deepmerge(result, variantStyle);
-                  });
-
-                  return result;
-                };
-              }
-              return transformedStylesArg;
-            }
-            return stylesArg;
-          })
-        : [];
-
-      let transformedStyleArg = styleArg;
-
-      if (isPlainObject(styleArg)) {
-        let styledArgVariants;
-        if (styleArg && styleArg.variants) {
-          styledArgVariants = styleArg.variants;
-          delete transformedStyleArg.variants;
-
-          transformedStyleArg = (props) => {
-            let result = styleArg;
-            const variantStyles = variantsResolver(
-              props,
-              transformVariants(styledArgVariants),
-              styledArgVariants,
-            );
-            variantStyles.forEach((variantStyle) => {
-              result = deepmerge(result, variantStyle);
-            });
-
-            return result;
-          };
-        }
-      } else if (
-        typeof styleArg === 'function' &&
-        // On the server Emotion doesn't use React.forwardRef for creating components, so the created
-        // component stays as a function. This condition makes sure that we do not interpolate functions
-        // which are basically components used as a selectors.
-        styleArg.__emotion_real !== styleArg
+
+    const transformStyleArg = (stylesArg) => {
+      // On the server Emotion doesn't use React.forwardRef for creating components, so the created
+      // component stays as a function. This condition makes sure that we do not interpolate functions
+      // which are basically components used as a selectors.
+      if (
+        (typeof stylesArg === 'function' && stylesArg.__emotion_real !== stylesArg) ||
+        isPlainObject(stylesArg)
       ) {
-        // If the type is function, we need to define the default theme.
-        transformedStyleArg = (props) =>
-          muiStyledFunctionResolver({ styledArg: styleArg, props, defaultTheme, themeId });
+        return (props) =>
+          processStyleArg(stylesArg, {
+            ...props,
+            theme: resolveTheme({ theme: props.theme, defaultTheme, themeId }),
+          });
       }
+      return stylesArg;
+    };
+    const muiStyledResolver = (styleArg, ...expressions) => {
+      let transformedStyleArg = transformStyleArg(styleArg);
+      const expressionsWithDefaultTheme = expressions ? expressions.map(transformStyleArg) : [];
 
       if (componentName && overridesResolver) {
         expressionsWithDefaultTheme.push((props) => {
           const theme = resolveTheme({ ...props, defaultTheme, themeId });
-          const styleOverrides = getStyleOverrides(componentName, theme);
```

Comment:
`getStyleOverrides` is not reused, so remove it to improve shifting between functions.